### PR TITLE
[CSE] Support GPU driver dedicated VHD

### PR DESCRIPTION
**Pull request by bowang-666**
CSE support GPU drivers and nvdia device plugin systemd pre-baked VHD

### CODE REVIEW

**Comment by xuto2** [comment]
@adilliadil PLAL

File: pkg/agent/utils.go [hunk 1]
```diff
@@ -333,6 +334,29 @@ func getDataDisks(a *api.AgentPoolProfile) string {
 	return buf.String()
 }
 
+func getNeedConfigureGPUDriver(properties *api.Properties) bool {
+	for _, agentPool := range properties.AgentPoolProfiles {
+		if agentPool.ImageRef != nil {
```
Comment (bowang-666) [inline pkg/agent/utils.go:339]: This will need to be updated to distro

Comment (bowang-666) [reply pkg/agent/utils.go]: changed to pass from RP, instead of relying on distro

File: parts/linux/cloud-init/artifacts/cse_cmd.sh [hunk 2]
```diff
@@ -56,4 +56,6 @@ IS_VHD={{GetVariable "isVHD"}}
 GPU_NODE={{GetVariable "gpuNode"}}
 SGX_NODE={{GetVariable "sgxNode"}}
 AUDITD_ENABLED={{GetVariable "auditdEnabled"}} 
+NEED_CONFIG_GPU_DRIVERS={{GetVariable "needConfigGPUDrivers"}} 
+ENABLE_GPU_SYSTEMD={{GetVariable "enableGPUSystemd"}}
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_cmd.sh:60]: ENABLE_GPU_DEVICE_PLUGIN_SYSTEMD

File: parts/linux/cloud-init/artifacts/cse_helpers.sh [hunk 3]
```diff
@@ -43,6 +43,8 @@ ERR_CONTAINERD_DOWNLOAD_TIMEOUT=70 {{/* Timeout waiting for containerd downloads
 ERR_CUSTOM_SEARCH_DOMAINS_FAIL=80 {{/* Unable to configure custom search domains */}}
 ERR_GPU_DRIVERS_START_FAIL=84 {{/* nvidia-modprobe could not be started by systemctl */}}
 ERR_GPU_DRIVERS_INSTALL_TIMEOUT=85 {{/* Timeout waiting for GPU drivers install */}}
+ERR_GPU_DEVICEPLUGIN_START_FAIL=86 {{/* nvidia device plugin could not be started by systemctl */}}
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_helpers.sh:46]: DEVICE**_**PLUGIN?

File: pkg/agent/variables.go [hunk 4]
```diff
@@ -44,7 +45,7 @@ func getCustomDataVariables(cs *api.ContainerService) paramsMap {
 }
 
 func getCSECommandVariables(cs *api.ContainerService, profile *api.AgentPoolProfile,
-	tenantID, subscriptionID, resourceGroupName, userAssignedIdentityID string) paramsMap {
+	tenantID, subscriptionID, resourceGroupName, userAssignedIdentityID string, needConfigGPUDrivers, enableGPUSystemd bool) paramsMap {
```
Comment (xuto2) [inline pkg/agent/variables.go:48]: nit: enableGPUDevicePluginSystemd

File: pkg/agent/variables.go [hunk 5]
```diff
@@ -69,6 +70,8 @@ func getCSECommandVariables(cs *api.ContainerService, profile *api.AgentPoolProf
 		"gpuNode":                         strconv.FormatBool(common.IsNvidiaEnabledSKU(profile.VMSize)),
 		"sgxNode":                         strconv.FormatBool(common.IsSgxEnabledSKU(profile.VMSize)),
 		"auditdEnabled":                   strconv.FormatBool(to.Bool(profile.AuditDEnabled)),
+		"needConfigGPUDrivers":            needConfigGPUDrivers,
+		"enableGPUSystemd":                enableGPUSystemd,
```
Comment (xuto2) [inline pkg/agent/variables.go:74]: nit: enableGPUDevicePluginSystemd

Comment (xuto2) [reply pkg/agent/variables.go:74]: or enableGPUDevicePlugin If the previous one is too long

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 6]
```diff
@@ -466,8 +466,32 @@ configGPUDrivers() {
...
+}
+
 ensureGPUDrivers() {
-    configGPUDrivers
+    if [[ "${NEED_CONFIG_GPU_DRIVERS}" = true ]]; then
+        echo "running configGPUDrivers +new"
+        configGPUDrivers
+    else
+        echo "skipping configGPUDrivers +new"
+        echo "start testGPUDrivers +new"
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_config.sh:492]: remove

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 7]
```diff
@@ -466,8 +466,32 @@ configGPUDrivers() {
...
+    fi
+}
+
 ensureGPUDrivers() {
-    configGPUDrivers
+    if [[ "${NEED_CONFIG_GPU_DRIVERS}" = true ]]; then
+        echo "running configGPUDrivers +new"
+        configGPUDrivers
+    else
+        echo "skipping configGPUDrivers +new"
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_config.sh:491]: update

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 8]
```diff
@@ -466,8 +466,32 @@ configGPUDrivers() {
...
+        fi
+    else
+        echo "gpu driver working fine"
+    fi
+}
+
 ensureGPUDrivers() {
-    configGPUDrivers
+    if [[ "${NEED_CONFIG_GPU_DRIVERS}" = true ]]; then
+        echo "running configGPUDrivers +new"
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_config.sh:488]: update

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 9]
```diff
@@ -466,8 +466,32 @@ configGPUDrivers() {
     retrycmd_if_failure 120 5 25 $GPU_DEST/bin/nvidia-smi || exit $ERR_GPU_DRIVERS_START_FAIL
     retrycmd_if_failure 120 5 25 ldconfig || exit $ERR_GPU_DRIVERS_START_FAIL
 }
+
+validateGPUDrivers() {
+    {{/* will install GPU drivers if not loaded correctly */}}
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_config.sh:471]: nit: remove this to minimize size of the cse payload as it's delivered through cloud-init that has a size limit of 64K

File: parts/linux/cloud-init/artifacts/cse_cmd.sh (hunk 2 above)
Comment (bowang-666) [reply parts/linux/cloud-init/artifacts/cse_cmd.sh]: changed to enableGPUDevicePlugin to make it consistent with the error code

File: parts/linux/cloud-init/artifacts/cse_config.sh [hunk 10]
```diff
@@ -466,8 +466,28 @@ configGPUDrivers() {
     retrycmd_if_failure 120 5 25 $GPU_DEST/bin/nvidia-smi || exit $ERR_GPU_DRIVERS_START_FAIL
     retrycmd_if_failure 120 5 25 ldconfig || exit $ERR_GPU_DRIVERS_START_FAIL
 }
+
+validateGPUDrivers() {
+    retrycmd_if_failure 24 5 25 nvidia-modprobe -u -c0 && echo "gpu driver loaded" || configGPUDrivers || exit $ERR_GPU_DRIVERS_START_FAIL
+    SMI_RESULT=$(retrycmd_if_failure 24 5 25 $GPU_DEST/bin/nvidia-smi)
+    SMI_STATUS=$?
+    if [[ $SMI_STATUS != 0 ]]; then
+        if [[ $SMI_RESULT == *"infoROM is corrupted"*  ]]; then
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/cse_config.sh:475]: extra space?

**Pull request merged**